src/CppAst.CodeGen/CSharp/CSharpElement.cs

```diff
@@ -9,9 +9,9 @@ namespace CppAst.CodeGen.CSharp
 {
     public abstract class CSharpElement : ICSharpElement
     {
-        public CppElement CppElement { get; internal set; }
+        public CppElement CppElement { get; set; }
 
-        public CSharpElement Parent { get; internal set; }
```

Comment:
One concern with making CSharpElement Parent public  is that it can go wrong unexpectedly with containers, as we are not protected well today. The only case this Parent is used is for containers... so do you really need this public?

Reply:
So its also set in the `DefaultParameterConverter`. Which is similar to `CppElement` if its going to be used in the default plugins and I want to be able to override or replace the default plugin I should be able to at least able replicate the behavior of the default plugin.

Reply:
Right, so that's a reason why it was not exposed.

Exposing it with public requires to protect all setters with a proper check of the parent (e.g like done in scriban [here](https://github.com/scriban/scriban/blob/6677f8f1537c23e994dd226b91b38c11e40e7291/src/Scriban/Syntax/Expressions/ScriptAssignExpression.cs#L32) and [ParentToThis](https://github.com/scriban/scriban/blob/6677f8f1537c23e994dd226b91b38c11e40e7291/src/Scriban/Syntax/ScriptNode.cs#L123-L136)).

It's a bit laborious, but that's quite mandatory to do when you expose re-parenting, otherwise it becomes quickly a nightmare to debug/investigate when there is a bug.

Reply:
I have no issue adding code similar to Scriban if that's what it would take to expose it. 

Because I just am not sure how folks are expected to us a Plugin API where I can't replicate the default behavior. Like for example if all I wanted to do was slightly change the ParameterConverter I can't.

Reply:
> Because I just am not sure how folks are expected to us a Plugin API where I can't replicate the default behavior. Like for example if all I wanted to do was slightly change the ParameterConverter I can't.

Yep, and that's ok. In the sense that I developed CppAst.CodeGen for my own use case. I remember that I made this decision for accessibility of CSharpElement, and that was precisely that I didn't have time to make it clean.  😉